Fix #122: Add FAN expansion mode

kiewietn opened the pull request:
<!--
Any HTML comment will be stripped when the markdown is rendered, so you don't need to delete them.
-->

<!-- You can remove this first section if you have contributed before -->
<!--
Any HTML comment will be stripped when the markdown is rendered, so you don't need to delete them.
-->

<!-- You can remove this first section if you have contributed before -->
### First time contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] I have read [how to contribute](/.github/CONTRIBUTING.md) to this project
- [x] I have read [the code of conduct](/.github/CODE_OF_CONDUCT.md) to this project

### Contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] I am using the provided [codeStyleConfig.xml](/.idea/codeStyles)
- [x] I have tested my contribution on these devices:
 * Huawei Y3 CAG-L02, Android 8.1.0
 * Samsung Galaxy S5 SM-G900H, Android 6.0.1
 * Virtual device Nexus 5, Android 8.1.0
- [x] My contribution is fully baked and ready to be merged as is
- [x] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit 
      using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-using-keywords/)

----------

### Description
<!--
Describe briefly what your pull request proposes to fix. Especially if you have more than one commit, it is helpful to 
give a summary of what your contribution as a whole is trying to solve.
Also, please describe shortly how you tested that your fix actually works.
-->
Add an expansion mode, where the items are arranged radially around the main FAB. 
Items with label text set, will have the labels display below the mini fab item.
The radius length of the expansion mode can be changed from the xml layout file using the `layout_constraintCircleRadius`. 

![fab_expansion](https://user-images.githubusercontent.com/31987564/62626239-3b398800-b927-11e9-8f7f-4873c06c7b0f.gif)

kiewietn committed:
Fix #122: Add FAN expansion mode for speed dial items.

Change FabWithLabelView to ConstraintLayout.
- This allows both horizontal and vertical expansion modes to show labels

Set layout params for each menu item to mimic expansion mode
for top/bottom/left and right

Expose hiding of label from FabWithLabelView
- For FAN expansion mode, the text is displayed below mini fab
- for right or left expansion modes, the text is always disabled. Same
as existing behaviour

Format code according to code style

Update sample app to show FAN expansion mode

kiewietn commented:
Not really sure how to handle the overlapping text. 
My thought was that clients will use shorter length labels, less action items or increase the expansion radius when text overlaps.

leinardi commented:
Hi @kiewietn, thanks a lot for the Pull Request, but I'm afraid I will not be able to accept it until the issue with the overlapping/cut labels is solved: I'm absolutely positive that, if the current implementation will be released, I would get many bug reports for this problem.

leinardi reviewed (changes requested):
Hi @kiewietn, thanks a lot for the Pull Request, but I'm afraid I will not be able to accept it until the issue with the overlapping/cut labels is solved: I'm absolutely positive that, if the current implementation will be released, I would get many bug reports for this problem.

kiewietn commented:
@leinardi, thanks for the comment.

To mitigate risk of overlapping labels, the maximum number of characters displayed on the label could be limited when using the FAN expansion mode.

The label length restrictions applied to the other expansion modes remain unchanged.

![fan_expand_text_overlap](https://user-images.githubusercontent.com/31987564/62857929-2da24a80-bcf9-11e9-8fc2-af77059696d9.gif)

Please let me know if this is satisfactory.

leinardi commented:
Hi @kiewietn, sorry for the late reply, busy week at work.

This is definitely an improvement but I'm still afraid that people will not understand this limitation and will keep opening issues complaining for the cut labels.

I would be fine to release this feature as experimental but I'm not sure how to do it with Java.
With Kotlin it would be possible to use the [Experimental API marker](https://kotlinlang.org/docs/reference/experimental.html).

I also notice a bug with the action items alignment:
![image](https://user-images.githubusercontent.com/273338/63081952-3edf9700-bf45-11e9-844f-e3b36e2600f0.png)

It seems that, on your fork, they are not aligned to the main FAB anymore:
![image](https://user-images.githubusercontent.com/273338/63082198-e3fa6f80-bf45-11e9-8c8a-0a917e57030e.png)

Upstream master doesn't have the issue:
![image](https://user-images.githubusercontent.com/273338/63082251-0096a780-bf46-11e9-9af9-da652b8c65be.png)

kiewietn committed:
Center horizontally on main fab when expanding TOP and BOTTOM

kiewietn committed:
Limit label length when expansion mode is FAN. This is to mitigate the
affect of long labels overlapping.

kiewietn commented:
Thanks for coming back to me @leinardi .
I have fixed the alignment issue and added the limit to the label lengths.

![Screenshot_1565944729](https://user-images.githubusercontent.com/31987564/63155049-27baab00-c012-11e9-83e3-9c47d3ef89ec.png)
